### ISSUE TITLE
Pylance crash after saving loose untitled notebook

### ISSUE THREAD

**Issue by debonte** [microsoft/pylance-release#4733]
## Environment data

-   Language Server version: 2023.8.31
-   OS and version: Windows 11
-   Python version: 3.11.4

## Repro Steps

1. Start VS Code. Do not open a folder/workspace.
2. `Create: New Jupyter Notebook`
3. Add two more blank code cells, adding each at the bottom of the notebook.
4. Save file
5. If Pylance doesn't crash, try adding another blank code cell at the bottom of the notebook.

## Expected behavior

No crash.

## Actual behavior

```text
2023-08-18 16:31:41.611 [info] Error: Debug Failure. False expression: Chained file path undefined doesn't match cellFilePaths c:\temp\temp\temp17.ipynb:pylance-notebook-cell:W0sZmlsZQ==.py
    at verifyCellChainIsLinear (c:\src\pyrx\packages\vscode-pylance\dist\server.bundle.js:23341:23)
    at updateCellChain (c:\src\pyrx\packages\vscode-pylance\dist\server.bundle.js:23428:5)
    at updateNotebookStructure (c:\src\pyrx\packages\vscode-pylance\dist\server.bundle.js:23436:5)
    at PylanceAsyncServer._onDidChangeNotebookDocumentAsync (c:\src\pyrx\packages\vscode-pylance\dist\server.bundle.js:22980:50)
```

[log.txt](https://github.com/microsoft/pylance-release/files/12384474/log.txt)

**Comment by debonte** [comment]
Found while investigating https://github.com/microsoft/pylance-release/issues/4685. I'm not sure if they are directly related.

The underlying problem here is that when `pythonPath` is an empty string,  `_getDefaultWorskpaceKey` searches for `<default>:mutable` but then creates an immutable workspace with key `<default>:`. Downstream when `_add` is called for the new workspace, the previous default workspace is deleted.

When Pylance crashes after saving the untitled notebook. the cause is that the workspace containing the first cell has been removed and therefore when `verifyCellChainIsLinear` calls `getChainedFilePath` for that cell, expecting the prefix cell path, it gets `undefined` instead.

**Comment by debonte** [comment]
I was able to hit the problem shown in this #4685 comment and confirmed that they are caused by the same issue: https://github.com/microsoft/pylance-release/issues/4685#issuecomment-1670070853

1. Start VS Code. Do not open a folder/workspace.
2. `Create: New Jupyter Notebook`
3. In first cell, type `thing = "thing"`
4. Add second cell below first, and add `thing`
5. Add third cell below second, and add `print(thing)`
6. Delete second (middle) cell.
7. Ctrl-Z to undo the delete.

When I undo the cell deletion, we get an `executeCodeAction` request which calls `getWorkspaceForFile` and that results in the `_getOrCreateDefaultWorkspace` + `_add` behavior described above which causes the workspace containing all of the existing cells to be removed from the workspace map.

We then get semantic token, inlay hint, and folding requests which also replace the default workspace although that hardly matters at that point.

If you hover over identifiers in the cells at this point to see if Pylance is still happy (as the user does in the video), hover fails because the cell files aren't in the workspace. Btw, hover also causes the default workspace to be replaced.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.8.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023841-23-August-2023-PreRelease)
